core/src/main/scala/magnolia1/magnolia.scala

```diff
@@ -393,6 +393,31 @@ object Magnolia {
           $impl
       """
 
+      def constructPartialAssignmentFunction(typeclasses: List[(Type, Tree)], arrayType: Type): (TermName, Tree) = {
+        val name = c.freshName(TermName("partialAssignments"))
+        val assignments = typeclasses.zipWithIndex.map { case ((subType, typeclass), idx) =>
+          val symbol = subType.typeSymbol
+          val (annotationTrees, inheritedAnnotationTrees) = annotationsOf(symbol)
+          q"""arr(start + $idx) = $SubtypeObj[$typeConstructor, $genericType, $subType](
+            ${typeNameOf(subType)},
+            $idx,
+            $ArrayObj(..${annotationTrees}): Array[_root_.scala.Any],
+            $ArrayObj(..${inheritedAnnotationTrees}): Array[_root_.scala.Any],
+            $ArrayObj(..${typeAnnotationsOf(symbol, fromParents = true)}): Array[_root_.scala.Any],
+            $CallByNeedObj($typeclass),
+            (t: $genericType) => t.isInstanceOf[$subType],
+            (t: $genericType) => t.asInstanceOf[$subType]
+          )"""
+        }
+        val tree =
+          q"""def $name(arr: Array[$arrayType], start: Int) = {
```

Comment:
Extract this to the top level:
```scala
val ArrayTpe = typeOf[Array[Any]].typeConstructor
```

And use it like this:
```suggestion
          q"""def $name(arr: $ArrayTpe[$arrayType], start: $IntTpe) = {
```

Reply:
[be8bd49](https://github.com/softwaremill/magnolia/pull/485/commits/be8bd496fe5891d6f7fea62ed31e3a3122cf2b8c)